cleanup_test.TestCleanup.test_drop_table_during_cleanup fails due to seastar::gate_closed_exception logged in logging message

we drop tables while performing cleanup to them in parallel. but there are chances that the operation hits a table whose `gate` is closed, and hence the error in logging message.

```
failed on teardown with "AssertionError: Unexpected errors found: [('node1', ['ERROR 2023-12-06 06:42:23,055 [shard 0:strm] api - force_keyspace_cleanup: keyspace=ks tables={table{name=cf2, id=91e455d0-9402-11ee-81cc-18cc756a1466}, table{name=cf1, id=88f5f550-9402-11ee-81cc-18cc756a1466}, table{name=cf0, id=80d60860-9402-11ee-81cc-18cc756a1466}} failed: seastar::gate_closed_exception (gate closed)', 'ERROR 2023-12-06 06:42:23,003 [shard 0:strm] task_manager - Failed force_keyspace_cleanup of ks.cf0: seastar::gate_closed_exception (gate closed)'])]"
...
```
see also https://jenkins.scylladb.com/job/scylla-master/job/next/6900/testReport/junit/cleanup_test/TestCleanup/Build___x86___dtest___test_drop_table_during_cleanup/

Similar on https://jenkins.scylladb.com/view/master/job/scylla-master/job/next/6919/testReport/junit/cleanup_test/TestCleanup/Build___x86___dtest___test_drop_table_during_cleanup_2/

`
failed on teardown with "AssertionError: Unexpected errors found: [('node1', ['ERROR 2023-12-10 12:41:49,093 [shard 0:strm] task_manager - Failed force_keyspace_cleanup of ks.cf0: seastar::gate_closed_exception (gate closed)', 'ERROR 2023-12-10 12:41:49,093 [shard 0:strm] api - force_keyspace_cleanup: keyspace=ks tables={table{name=cf2, id=7229f560-9759-11ee-b1c9-99c34ed95232}, table{name=cf1, id=68fb5790-9759-11ee-b1c9-99c34ed95232}, table{name=cf0, id=60d17f90-9759-11ee-b1c9-99c34ed95232}} failed: seastar::gate_closed_exception (gate closed)', 'ERROR 2023-12-10 12:41:49,093 [shard 1:strm] task_manager - Failed force_keyspace_cleanup of ks.cf0: seastar::gate_closed_exception (gate closed)'])]"
`

--- a/compaction/task_manager_module.cc
+++ b/compaction/task_manager_module.cc
@@ -198,6 +198,8 @@ future<> run_on_table(sstring op, replica::database& db, std::string keyspace, t
         co_await func(t);
     } catch (const replica::no_such_column_family& e) {
         tasks::tmlogger.warn("Skipping {} of {}.{}: {}", op, keyspace, ti.name, e.what());
+    } catch (const gate_closed_exception& e) {
+        tasks::tmlogger.warn("Skipping {} of {}.{}: {}", op, keyspace, ti.name, e.what());
     } catch (...) {
         ex = std::current_exception();
         tasks::tmlogger.error("Failed {} of {}.{}: {}", op, keyspace, ti.name, ex);

--- a/compaction/task_manager_module.cc
+++ b/compaction/task_manager_module.cc
@@ -193,7 +193,9 @@ future<> run_on_table(sstring op, replica::database& db, std::string keyspace, t
     std::exception_ptr ex;
     tasks::tmlogger.debug("Starting {} on {}.{}", op, keyspace, ti.name);
     try {
-        co_await func(db.find_column_family(ti.id));
+        auto& t = db.find_column_family(ti.id);
+        auto holder = t.hold();
+        co_await func(t);
     } catch (const replica::no_such_column_family& e) {
         tasks::tmlogger.warn("Skipping {} of {}.{}: {}", op, keyspace, ti.name, e.what());
     } catch (...) {

--- a/replica/database.hh
+++ b/replica/database.hh
@@ -361,6 +361,31 @@ struct table_stats {
 
 using storage_options = data_dictionary::storage_options;
 
+// Smart table pointer that guards the table object
+// while it's being accessed asynchronously
+class table_holder {
+    gate::holder _holder;
+    lw_shared_ptr<table> _table_ptr;
+public:
+    explicit table_holder(table&);
+
+    const table& operator*() const noexcept {
+        return *_table_ptr;
+    }
+
+    table& operator*() noexcept {
+        return *_table_ptr;
+    }
+
+    const table* operator->() const noexcept {
+        return _table_ptr.operator->();
+    }
+
+    table* operator->() noexcept {
+        return _table_ptr.operator->();
+    }
+};
+
 class table : public enable_lw_shared_from_this<table>
             , public weakly_referencable<table> {
 public:
@@ -783,6 +808,11 @@ public:
 
     table(column_family&&) = delete; // 'this' is being captured during construction
     ~table();
+
+    table_holder hold() {
+        return table_holder(*this);
+    }
+
     const schema_ptr& schema() const { return _schema; }
     void set_schema(schema_ptr);
     db::commitlog* commitlog() const;

--- a/replica/table.cc
+++ b/replica/table.cc
@@ -67,6 +67,11 @@ static seastar::metrics::label keyspace_label("ks");
 
 using namespace std::chrono_literals;
 
+table_holder::table_holder(table& t)
+    : _holder(t.async_gate())
+    , _table_ptr(t.shared_from_this())
+{ }
+
 void table::update_sstables_known_generation(sstables::generation_type generation) {
     auto gen = generation ? generation.as_int() : 0;
     if (_sstable_generation_generator) {

--- a/replica/table.cc
+++ b/replica/table.cc
@@ -1073,10 +1073,13 @@ table::stop() {
     if (_async_gate.is_closed()) {
         co_return;
     }
-    co_await _async_gate.close();
+    // Allow `compaction_group::stop` to stop ongoing compactions
+    // while they may still hold the table _async_gate
+    auto gate_closed_fut = _async_gate.close();
     co_await await_pending_ops();
     co_await parallel_foreach_compaction_group(std::mem_fn(&compaction_group::stop));
     co_await _sstable_deletion_gate.close();
+    co_await std::move(gate_closed_fut);
     co_await get_row_cache().invalidate(row_cache::external_updater([this] {
         for (const compaction_group_ptr& cg : compaction_groups()) {
             cg->clear_sstables();

